### PR TITLE
8327661: C1: Make RBP allocatable on x64 when PreserveFramePointer is disabled

### PR DESCRIPTION
**Pull request by D-D-H**
Hi,

Could I have a review of this change that makes RBP allocatable in c1 register allocation when PreserveFramePointer is not enabled.

There seems no reason that RBP cannot be used. Although the performance of c1 jit code is not very critical, in my opinion, this change will not add overhead of compilation. So maybe it is acceptable.

I am not very sure if I have changed all the places that should be.

Testing: fastdebug tier1-4 on Linux x64

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8327661](https://bugs.openjdk.org/browse/JDK-8327661): C1: Make RBP allocatable on x64 when PreserveFramePointer is disabled (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18167/head:pull/18167` \
`$ git checkout pull/18167`

Update a local copy of the PR: \
`$ git checkout pull/18167` \
`$ git pull https://git.openjdk.org/jdk.git pull/18167/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18167`

View PR using the GUI difftool: \
`$ git pr show -t 18167`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18167.diff">https://git.openjdk.org/jdk/pull/18167.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18167#issuecomment-1988207618)

### CODE REVIEW

File: src/hotspot/cpu/x86/c1_Defs_x86.hpp [hunk 1]
```diff
@@ -44,7 +44,7 @@ enum {
   pd_nof_xmm_regs_frame_map = XMMRegister::number_of_registers,    // number of registers used during code emission
 
 #ifdef _LP64
-  #define UNALLOCATED 4    // rsp, rbp, r15, r10
+  #define UNALLOCATED 3    // rsp, r15, r10
```
Comment (dean-long) [inline src/hotspot/cpu/x86/c1_Defs_x86.hpp:47]: This affects pd_nof_caller_save_cpu_regs_frame_map below, but RBP is callee-saved, not caller-saved.

Comment (D-D-H) [reply src/hotspot/cpu/x86/c1_Defs_x86.hpp:47]: Yes. I updated the patch.

I want to confirm, if we treat RBP as caller saved, is there any correctness problem?

File: src/hotspot/share/c1/c1_LinearScan.cpp [hunk 2]
```diff
@@ -5752,6 +5752,7 @@ void LinearScanWalker::alloc_locked_reg(Interval* cur) {
 
 bool LinearScanWalker::no_allocation_possible(Interval* cur) {
 #ifdef X86
+#ifndef _LP64
```
Comment (D-D-H) [inline src/hotspot/share/c1/c1_LinearScan.cpp:5755]: rbp is callee-saved, so the following logic doesn't work. That'll slow down the allocation.

**Comment by D-D-H** [comment]
Gentle ping.

Since the benefits are not obvious, I'll close this PR if there are no reviews for one more week.

**Comment by dean-long** [comment]
Yes, the risk / reward ratio on this seems borderline.  Even of arm32 or x86_32, where there are fewer registers, it's not clear how much using FP in C1 would help.

**Pull request closed without merging**